File: apps/web/lib/api/accounting.ts
import { publicEnv } from "@/lib/config";
import { safeApiErrorMessage } from "@/lib/api/safe-error";

export type Company = {
  id: string;
  legal_name: string;
  trade_name: string | null;
  gstin: string | null;
  state_code: string | null;
};

export type DemoCompanyResponse = {
  company_id: string;
  legal_name: string;
  seeded_ledgers: number;
  seeded_vouchers: number;
  seeded_invoices: number;
  seeded_bank_transactions: number;
};

export type ClientDemoWorkspaceResponse = {
  success: true;
  mode: "client_demo";
  company_id: string;
  company_name: string;
  user: {
    name: string;
    email: string;
    role: "owner";
  };
};

export type CompanyCreate = {
  legal_name: string;
  trade_name?: string | null;
  gstin?: string | null;
  state_code?: string | null;
};

export type AccessRequest = {
  id: string;
  company_id: string;
  company_legal_name: string;
  requester_profile_id: string;
  requester_email: string | null;
  requested_role: "accountant" | "viewer";
  status: "pending" | "approved" | "rejected";
  created_at: string;
  decided_at: string | null;
};

export type AccountNature = "asset" | "liability" | "equity" | "income" | "expense";
export type LedgerCategory =
  | "cash"
  | "bank"
  | "sundry_debtor"
  | "sundry_creditor"
  | "sales"
  | "purchase"
  | "direct_expense"
  | "indirect_expense"
  | "direct_income"
  | "indirect_income"
  | "input_gst"
  | "output_gst"
  | "round_off"
  | "capital"
  | "loan"
  | "other";

export type LedgerGroup = {
  id: string;
  name: string;
  account_nature: AccountNature;
  parent_id: string | null;
  is_system: boolean;
};

export type Ledger = {
  id: string;
  name: string;
  ledger_group_id: string;
  group_name: string;
  category: LedgerCategory;
  account_nature: AccountNature;
  opening_balance: string;
  opening_balance_type: "dr" | "cr";
  gstin: string | null;
  state_code: string | null;
  is_system: boolean;
  is_active: boolean;
};

export type VoucherType =
  | "receipt"
  | "payment"
  | "contra"
  | "journal"
  | "purchase"
  | "sales"
  | "debit_note"
  | "credit_note";

export type Voucher = {
  id: string;
  voucher_number: string;
  voucher_type: VoucherType;
  voucher_date: string;
  status: string;
  narration: string | null;
  posted_at: string | null;
  lines: Array<{
    id: string;
    ledger_id: string;
    ledger_name: string;
    debit: string;
    credit: string;
    narration: string | null;
  }>;
};

export type AuditEvent = {
  id: string;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string | null;
  action_type: string;
  entity_type: string;
  entity_id: string;
  summary: string;
};

export type DashboardMetrics = {
  revenue: string;
  expenses: string;
  profit: string;
  cash_position: string;
  receivables: string;
  payables: string;
};

export type DebugCounts = {
  ledgers: number;
  vouchers: number;
  voucher_lines: number;
  accounting_entries: number;
  invoices: number;
  invoice_items: number;
  bank_transactions: number;
  audit_logs: number;
  ai_logs: number;
  document_ai_logs: number;
  inventory_items: number;
};

export type TrialBalanceRow = {
  ledger_id: string;
  ledger_name: string;
  account_nature: AccountNature;
  category: LedgerCategory;
  debit: string;
  credit: string;
};

export type Invoice = {
  id: string;
  invoice_type: "sales" | "purchase";
  invoice_number: string;
  invoice_date: string;
  due_date: string | null;
  party_ledger_id: string;
  party_ledger_name: string | null;
  voucher_id: string | null;
  taxable_value: string;
  cgst_amount: string;
  sgst_amount: string;
  igst_amount: string;
  total_amount: string;
  notes: string | null;
  lines: Array<{
    id: string;
    description: string;
    hsn_sac: string | null;
    quantity: string;
    unit: string;
    unit_price: string;
    gst_rate: string;
    taxable_value: string;
    cgst_amount: string;
    sgst_amount: string;
    igst_amount: string;
    total_amount: string;
  }>;
};

export type InvoiceGstSummaryRow = {
  invoice_id: string;
  invoice_number: string;
  invoice_type: "sales" | "purchase";
  invoice_date: string;
  party_ledger_name: string;
  taxable_value: string;
  cgst_amount: string;
  sgst_amount: string;
  igst_amount: string;
  total_amount: string;
};

export type GstReport = {
  input_gst: string;
  output_gst: string;
  net_payable: string;
};

export type LedgerScrutinyIssue = {
  severity: "high" | "warning" | "info" | string;
  title: string;
  detail: string;
  amount: string | null;
};

export type LedgerScrutiny = {
  issue_count: number;
  high_risk_count: number;
  warning_count: number;
  issues: LedgerScrutinyIssue[];
};

export type TdsCalculatorResult = {
  taxable_amount: string;
  rate_percent: string;
  tds_amount: string;
  net_payable: string;
};

export type PfCalculatorResult = {
  eligible_wage: string;
  employee_contribution: string;
  employer_contribution: string;
  total_contribution: string;
};

export type EsicCalculatorResult = {
  eligible: boolean;
  eligible_wage: string;
  employee_contribution: string;
  employer_contribution: string;
  total_contribution: string;
};

type ApiOptions = {
  token: string;
  method?: "GET" | "POST" | "PATCH" | "DELETE";
  body?: unknown;
};

async function api<T>(path: string, options: ApiOptions): Promise<T> {
  const url = `${publicEnv.NEXT_PUBLIC_API_URL}${path}`;
  let response: Response;
  try {
    response = await fetch(url, {
      method: options.method ?? "GET",
      headers: {
        Authorization: `Bearer ${options.token}`,
        "Content-Type": "application/json"
      },
      body: options.body ? JSON.stringify(options.body) : undefined
    });
  } catch {
    throw new Error("ABHAY Intelligence is syncing. Your dashboard remains available.");
  }

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: "Request failed" }));
    throw new Error(safeApiErrorMessage(error.detail, "Accounting request could not be completed. Please try again."));
  }

  if (response.status === 204) {
    return undefined as T;
  }

  return response.json() as Promise<T>;
}

async function publicApi<T>(path: string, options: { method?: "GET" | "POST"; body?: unknown } = {}): Promise<T> {
  const url = `${publicEnv.NEXT_PUBLIC_API_URL}${path}`;
  let response: Response;
  try {
    response = await fetch(url, {
      method: options.method ?? "GET",
      headers: {
        "Content-Type": "application/json"
      },
      body: options.body ? JSON.stringify(options.body) : undefined
    });
  } catch {
    throw new Error("ABHAY backend is temporarily unavailable.");
  }

  const data = await response.json().catch(() => ({ detail: "Request failed" }));
  if (!response.ok) {
    throw new Error(safeApiErrorMessage(data.detail, "Request could not be completed. Please try again."));
  }
  return data as T;
}

export const accountingApi = {
  companies: (token: string) => api<Company[]>("/companies", { token }),
  createCompany: (token: string, body: CompanyCreate) =>
    api<Company>("/companies", { token, method: "POST", body }),
  requestAccess: (companyId: string, token: string, requestedRole: "accountant" | "viewer") =>
    api<AccessRequest>(`/companies/${companyId}/access-requests`, {
      token,
      method: "POST",
      body: { requested_role: requestedRole }
    }),
  accessRequests: (companyId: string, token: string) =>
    api<AccessRequest[]>(`/companies/${companyId}/access-requests`, { token }),
  decideAccessRequest: (
    companyId: string,
    requestId: string,
    token: string,
    decision: "approve" | "reject",
    role: "accountant" | "viewer"
  ) =>
    api<AccessRequest>(`/companies/${companyId}/access-requests/${requestId}`, {
      token,
      method: "PATCH",
      body: { decision, role }
    }),
  createDemoCompany: (token: string) =>
    api<DemoCompanyResponse>("/demo/company", { token, method: "POST" }),
  clientDemoWorkspace: () =>
    publicApi<ClientDemoWorkspaceResponse>("/api/demo/client-workspace", { method: "POST" }),
  groups: (companyId: string, token: string) =>
    api<LedgerGroup[]>(`/companies/${companyId}/ledger-groups`, { token }),
  createGroup: (companyId: string, token: string, body: { name: string; account_nature: AccountNature }) =>
    api<LedgerGroup>(`/companies/${companyId}/ledger-groups`, { token, method: "POST", body }),
  ledgers: (companyId: string, token: string, query = "") =>
    api<Ledger[]>(`/companies/${companyId}/ledgers${query}`, { token }),
  createLedger: (companyId: string, token: string, body: unknown) =>
    api<Ledger>(`/companies/${companyId}/ledgers`, { token, method: "POST", body }),
  updateLedger: (companyId: string, ledgerId: string, token: string, body: unknown) =>
    api<Ledger>(`/companies/${companyId}/ledgers/${ledgerId}`, { token, method: "PATCH", body }),
  deleteLedger: (companyId: string, ledgerId: string, token: string) =>
    api<void>(`/companies/${companyId}/ledgers/${ledgerId}`, { token, method: "DELETE" }),
  vouchers: (companyId: string, token: string) =>
    api<Voucher[]>(`/companies/${companyId}/vouchers`, { token }),
  createVoucher: (companyId: string, token: string, body: unknown) =>
    api<Voucher>(`/companies/${companyId}/vouchers`, { token, method: "POST", body }),
  auditEvents: (companyId: string, token: string) =>
    api<AuditEvent[]>(`/companies/${companyId}/audit-events`, { token }),
  dashboard: (companyId: string, token: string) =>
    api<DashboardMetrics>(`/companies/${companyId}/reports/dashboard`, { token }),
  debugCounts: (companyId: string, token: string) =>
    api<DebugCounts>(`/companies/${companyId}/debug-counts`, { token }),
  trialBalance: (companyId: string, token: string) =>
    api<TrialBalanceRow[]>(`/companies/${companyId}/reports/trial-balance`, { token }),
  profitAndLoss: (companyId: string, token: string) =>
    api<{ revenue: string; expenses: string; profit: string }>(
      `/companies/${companyId}/reports/profit-and-loss`,
      { token }
    ),
  balanceSheet: (companyId: string, token: string) =>
    api<{ assets: string; liabilities: string; equity: string; check_difference: string }>(
      `/companies/${companyId}/reports/balance-sheet`,
      { token }
    ),
  cashFlow: (companyId: string, token: string) =>
    api<{
      operating_cash_flow: string;
      investing_cash_flow: string;
      financing_cash_flow: string;
      net_cash_flow: string;
    }>(`/companies/${companyId}/reports/cash-flow`, { token }),
  invoices: (companyId: string, token: string) =>
    api<Invoice[]>(`/companies/${companyId}/invoices`, { token }),
  invoice: (companyId: string, invoiceId: string, token: string) =>
    api<Invoice>(`/companies/${companyId}/invoices/${invoiceId}`, { token }),
  createInvoice: (companyId: string, token: string, body: unknown) =>
    api<Invoice>(`/companies/${companyId}/invoices`, { token, method: "POST", body }),
  gstReport: (companyId: string, token: string) =>
    api<GstReport>(`/companies/${companyId}/reports/gst`, { token }),
  invoiceGstSummary: (companyId: string, token: string) =>
    api<InvoiceGstSummaryRow[]>(`/companies/${companyId}/reports/gst/invoices`, { token }),
  ledgerScrutiny: (companyId: string, token: string) =>
    api<LedgerScrutiny>(`/companies/${companyId}/reports/ledger-scrutiny`, { token }),
  calculateTds: (token: string, body: { amount: string; rate_percent: string }) =>
    api<TdsCalculatorResult>("/calculators/tds", { token, method: "POST", body }),
  calculatePf: (
    token: string,
    body: {
      monthly_basic_wage: string;
      employee_rate_percent: string;
      employer_rate_percent: string;
      wage_ceiling: string;
    }
  ) => api<PfCalculatorResult>("/calculators/pf", { token, method: "POST", body }),
  calculateEsic: (
    token: string,
    body: {
      monthly_gross_wage: string;
      employee_rate_percent: string;
      employer_rate_percent: string;
      wage_limit: string;
    }
  ) => api<EsicCalculatorResult>("/calculators/esic", { token, method: "POST", body }),
  gstr1CsvUrl: (companyId: string) =>
    `/api/reports/gstr1.csv?company_id=${encodeURIComponent(companyId)}`,
  gstr3bCsvUrl: (companyId: string) =>
    `/api/reports/gstr3b.csv?company_id=${encodeURIComponent(companyId)}`,
  invoicePdfUrl: (companyId: string, invoiceId: string) =>
    `/api/invoices/pdf?company_id=${encodeURIComponent(companyId)}&invoice_id=${encodeURIComponent(invoiceId)}`
};

File: apps/api/app/schemas/accounting.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.models.accounting import AccountNature, GstSupplyType, InvoiceType, LedgerCategory, VoucherType


class CompanyResponse(BaseModel):
    id: UUID
    legal_name: str
    trade_name: str | None = None
    gstin: str | None = None
    state_code: str | None = None

    model_config = {"from_attributes": True}


class CompanyCreate(BaseModel):
    legal_name: str = Field(min_length=2, max_length=180)
    trade_name: str | None = Field(default=None, max_length=180)
    gstin: str | None = Field(default=None, min_length=15, max_length=15)
    state_code: str | None = Field(default=None, min_length=2, max_length=2)


class AccessRequestCreate(BaseModel):
    requested_role: str = Field(default="accountant", pattern="^(accountant|viewer)$")


class AccessRequestDecision(BaseModel):
    decision: str = Field(pattern="^(approve|reject)$")
    role: str = Field(default="accountant", pattern="^(accountant|viewer)$")


class AccessRequestResponse(BaseModel):
    id: UUID
    company_id: UUID
    company_legal_name: str
    requester_profile_id: UUID
    requester_email: str | None = None
    requested_role: str
    status: str
    created_at: datetime
    decided_at: datetime | None = None


class DemoCompanyResponse(BaseModel):
    company_id: UUID
    legal_name: str
    seeded_ledgers: int
    seeded_vouchers: int
    seeded_invoices: int
    seeded_bank_transactions: int


class DebugCountsResponse(BaseModel):
    ledgers: int
    vouchers: int
    voucher_lines: int
    accounting_entries: int
    invoices: int
    invoice_items: int
    bank_transactions: int
    ai_logs: int
    document_ai_logs: int
    audit_logs: int
    inventory_items: int


class LedgerGroupCreate(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    account_nature: AccountNature
    parent_id: UUID | None = None


class LedgerGroupResponse(LedgerGroupCreate):
    id: UUID
    is_system: bool

    model_config = {"from_attributes": True}


class LedgerCreate(BaseModel):
    name: str = Field(min_length=2, max_length=160)
    ledger_group_id: UUID
    category: LedgerCategory = LedgerCategory.other
    account_nature: AccountNature
    opening_balance: Decimal = Decimal("0.00")
    opening_balance_type: str = Field(default="dr", pattern="^(dr|cr)$")
    gstin: str | None = Field(default=None, min_length=15, max_length=15)
    state_code: str | None = Field(default=None, min_length=2, max_length=2)


class LedgerUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=160)
    ledger_group_id: UUID | None = None
    category: LedgerCategory | None = None
    account_nature: AccountNature | None = None
    opening_balance: Decimal | None = None
    opening_balance_type: str | None = Field(default=None, pattern="^(dr|cr)$")
    gstin: str | None = Field(default=None, min_length=15, max_length=15)
    state_code: str | None = Field(default=None, min_length=2, max_length=2)
    is_active: bool | None = None


class LedgerResponse(BaseModel):
    id: UUID
    name: str
    ledger_group_id: UUID
    group_name: str
    category: LedgerCategory
    account_nature: AccountNature
    opening_balance: Decimal
    opening_balance_type: str
    gstin: str | None
    state_code: str | None
    is_system: bool
    is_active: bool


class VoucherLineCreate(BaseModel):
    ledger_id: UUID
    debit: Decimal = Decimal("0.00")
    credit: Decimal = Decimal("0.00")
    narration: str | None = None

    @field_validator("debit", "credit")
    @classmethod
    def amount_cannot_be_negative(cls, value: Decimal) -> Decimal:
        if value < 0:
            raise ValueError("Amount cannot be negative")
        return value


class VoucherCreate(BaseModel):
    voucher_type: VoucherType
    voucher_date: date
    narration: str | None = None
    lines: list[VoucherLineCreate] = Field(min_length=2)


class VoucherLineResponse(BaseModel):
    id: UUID
    ledger_id: UUID
    ledger_name: str
    debit: Decimal
    credit: Decimal
    narration: str | None


class VoucherResponse(BaseModel):
    id: UUID
    voucher_number: str
    voucher_type: VoucherType
    voucher_date: date
    status: str
    narration: str | None
    posted_at: datetime | None
    lines: list[VoucherLineResponse]


class AuditEventResponse(BaseModel):
    id: UUID
    created_by: UUID | None = None
    updated_by: UUID | None = None
    created_at: datetime
    updated_at: datetime | None = None
    action_type: str
    entity_type: str
    entity_id: UUID
    summary: str


class TrialBalanceRow(BaseModel):
    ledger_id: UUID
    ledger_name: str
    account_nature: AccountNature
    category: LedgerCategory
    debit: Decimal
    credit: Decimal


class ProfitAndLossResponse(BaseModel):
    revenue: Decimal
    expenses: Decimal
    profit: Decimal


class BalanceSheetResponse(BaseModel):
    assets: Decimal
    liabilities: Decimal
    equity: Decimal
    check_difference: Decimal


class CashFlowResponse(BaseModel):
    operating_cash_flow: Decimal
    investing_cash_flow: Decimal
    financing_cash_flow: Decimal
    net_cash_flow: Decimal


class DashboardMetrics(BaseModel):
    revenue: Decimal
    expenses: Decimal
    profit: Decimal
    cash_position: Decimal
    receivables: Decimal
    payables: Decimal


class InvoiceLineCreate(BaseModel):
    description: str = Field(min_length=2, max_length=240)
    hsn_sac: str | None = None
    quantity: Decimal = Field(gt=0)
    unit: str = "NOS"
    unit_price: Decimal = Field(ge=0)
    discount_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    gst_rate: Decimal = Field(default=Decimal("18.00"), ge=0)


class InvoiceCreate(BaseModel):
    invoice_type: InvoiceType
    invoice_number: str = Field(min_length=1, max_length=80)
    invoice_date: date
    due_date: date | None = None
    party_ledger_id: UUID
    gst_supply_type: GstSupplyType
    notes: str | None = None
    lines: list[InvoiceLineCreate] = Field(min_length=1)


class InvoiceLineResponse(BaseModel):
    id: UUID
    description: str
    hsn_sac: str | None
    quantity: Decimal
    unit: str
    unit_price: Decimal
    gst_rate: Decimal
    taxable_value: Decimal
    cgst_amount: Decimal
    sgst_amount: Decimal
    igst_amount: Decimal
    total_amount: Decimal


class InvoiceResponse(BaseModel):
    id: UUID
    invoice_type: InvoiceType
    invoice_number: str
    invoice_date: date
    due_date: date | None = None
    party_ledger_id: UUID
    party_ledger_name: str | None = None
    voucher_id: UUID | None = None
    taxable_value: Decimal
    cgst_amount: Decimal
    sgst_amount: Decimal
    igst_amount: Decimal
    total_amount: Decimal
    notes: str | None = None
    lines: list[InvoiceLineResponse]


class GstRateResponse(BaseModel):
    label: str
    rate: Decimal


class GstReportResponse(BaseModel):
    input_gst: Decimal
    output_gst: Decimal
    net_payable: Decimal


class InvoiceGstSummaryRow(BaseModel):
    invoice_id: UUID
    invoice_number: str
    invoice_type: InvoiceType
    invoice_date: date
    party_ledger_name: str
    taxable_value: Decimal
    cgst_amount: Decimal
    sgst_amount: Decimal
    igst_amount: Decimal
    total_amount: Decimal


class TdsCalculatorRequest(BaseModel):
    amount: Decimal = Field(ge=0)
    rate_percent: Decimal = Field(default=Decimal("10.00"), ge=0)


class TdsCalculatorResponse(BaseModel):
    taxable_amount: Decimal
    rate_percent: Decimal
    tds_amount: Decimal
    net_payable: Decimal


class PfCalculatorRequest(BaseModel):
    monthly_basic_wage: Decimal = Field(ge=0)
    employee_rate_percent: Decimal = Field(default=Decimal("12.00"), ge=0)
    employer_rate_percent: Decimal = Field(default=Decimal("12.00"), ge=0)
    wage_ceiling: Decimal = Field(default=Decimal("15000.00"), ge=0)


class PfCalculatorResponse(BaseModel):
    eligible_wage: Decimal
    employee_contribution: Decimal
    employer_contribution: Decimal
    total_contribution: Decimal


class EsicCalculatorRequest(BaseModel):
    monthly_gross_wage: Decimal = Field(ge=0)
    employee_rate_percent: Decimal = Field(default=Decimal("0.75"), ge=0)
    employer_rate_percent: Decimal = Field(default=Decimal("3.25"), ge=0)
    wage_limit: Decimal = Field(default=Decimal("21000.00"), ge=0)


class EsicCalculatorResponse(BaseModel):
    eligible: bool
    eligible_wage: Decimal
    employee_contribution: Decimal
    employer_contribution: Decimal
    total_contribution: Decimal


class LedgerScrutinyIssue(BaseModel):
    severity: str
    title: str
    detail: str
    amount: Decimal | None = None


class LedgerScrutinyResponse(BaseModel):
    issue_count: int
    high_risk_count: int
    warning_count: int
    issues: list[LedgerScrutinyIssue]
